go-controller/pkg/ovn/controller/services/load_balancer.go

```diff
@@ -0,0 +1,456 @@
+package services
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	globalconfig "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	ovnlb "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/loadbalancer"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+
+	v1 "k8s.io/api/core/v1"
+	discovery "k8s.io/api/discovery/v1beta1"
+	"k8s.io/klog/v2"
+)
+
+// lbConfig is the abstract desired load balancer configuration.
+// vips and endpoints are mixed families.
+type lbConfig struct {
+	vips     []string    // just ip or the special value "node" for the node's physical IPs (i.e. NodePort)
+	protocol v1.Protocol // TCP, UDP, or SCTP
+	inport   int32       // the incoming (virtual) port number
+	eps      util.LbEndpoints
+
+	routerLocalMode bool // if true, then vips added on the router are in "local" mode (used for ExternalTrafficPolicy)
+}
+
+// just used for consistent ordering
+var protos = []v1.Protocol{
+	v1.ProtocolTCP,
+	v1.ProtocolUDP,
+	v1.ProtocolSCTP,
+}
+
+// buildServiceLBConfigs generates the abstract load balancer(s) configurations for each service. The abstract configurations
+// are then expanded in buildClusterLBs and buildPerNodeLBs to the full list of OVN LBs desired.
+//
+// It creates two lists of configurations:
+// - the per-node configs, which are load balancers that, for some reason must be expanded per-node. (see below for why)
+// - the cluster-wide configs, which are load balancers that can be the same across the whole cluster.
+//
+// For a "standard" ClusterIP service (possibly with ExternalIPS or external LoadBalancer Status IPs),
+// a single cluster-wide LB will be created.
+//
+// Per-node LBs will be created for
+// - services with NodePort set
+// - services with host-network endpoints (for shared gateway mode)
+// - services with ExternalTrafficPolicy=Local
+func buildServiceLBConfigs(service *v1.Service, endpointSlices []*discovery.EndpointSlice) (perNodeConfigs []lbConfig, clusterConfigs []lbConfig) {
+	// For each svcPort, determine if it will be applied per-node or cluster-wide
```

Comment:
Determining if it will be applied per node or cluster wide isn't different for multiple svcPort's right? We determine if 
ALL of a service's loadbalancers should be applied per-node or cluster wide, and then need to build lbConfig's for each svcPort.

Reply:
Kind of. NodePort is a per-ServicePort property. ExternalIPs and host-network pods is per-Service. So it's a mix.

Theoretically we could have per-port Endpoints, but currently we do not.